java/compaction/compaction-core/src/main/java/sleeper/compaction/strategy/CompactionStrategyIndex.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright 2022-2024 Crown Copyright
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package sleeper.compaction.strategy;
+
+import sleeper.core.partition.Partition;
+import sleeper.core.statestore.FileReference;
+import sleeper.core.table.TableStatus;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+public class CompactionStrategyIndex {
+    private final List<FilesInPartition> filesInLeafPartitions;
+
+    public CompactionStrategyIndex(TableStatus tableStatus, List<FileReference> allFileReferences, List<Partition> allPartitions) {
+        Map<String, FilesInAscendingOrder> filesWithNoJobIdByPartition = new HashMap<>();
+        Map<String, List<FileReference>> filesWithJobIdByPartition = new HashMap<>();
+        List<String> leafPartitionIds = allPartitions.stream()
+                .filter(Partition::isLeafPartition)
+                .map(Partition::getId)
+                .collect(Collectors.toList());
+
+        allFileReferences.stream()
+                .filter(file -> leafPartitionIds.contains(file.getPartitionId()))
```

Comment:
Since leafPartitionIds was changed from a set to a list, this will loop through potentially all the leaf partitions for every file reference.

I can see a few options to avoid that:

- Maintain the ordering of the partitions in the set, e.g. with a linked hash set
- Handle the list of leaf partitions and the lookup for which partitions are leaves separately
- Don't filter the files at this point, only filter to the leaf partitions when you iterate through the partitions
- Redesign the algorithm some other way